src/core/lib/channel/metrics.h

```diff
@@ -0,0 +1,217 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_SRC_CORE_LIB_CHANNEL_METRICS_H
+#define GRPC_SRC_CORE_LIB_CHANNEL_METRICS_H
+
+#include <grpc/support/port_platform.h>
+
+#include <atomic>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/support/log.h>
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc_core {
+
+// Static only.
+class GlobalInstrumentsRegistry {
+ public:
+  enum class ValueType {
+    kUndefined,
+    kUInt64,
+    kDouble,
+  };
+  enum class InstrumentType {
+    kUndefined,
+    kCounter,
+    kHistogram,
+  };
+  struct GlobalInstrumentDescriptor {
+    ValueType value_type;
+    InstrumentType instrument_type;
+    uint32_t index;
+    absl::string_view name;
+    absl::string_view description;
+    absl::string_view unit;
+    std::vector<absl::string_view> label_keys;
+    std::vector<absl::string_view> optional_label_keys;
+  };
+  struct GlobalHandle {
+    // This is the index for the corresponding registered instrument that
+    // StatsPlugins can use to uniquely identify an instrument in the current
+    // process. Though this is not guaranteed to be stable between different
+    // runs or between different versions.
+    uint32_t index;
+  };
+  struct GlobalUInt64CounterHandle : GlobalHandle {};
+  struct GlobalDoubleCounterHandle : GlobalHandle {};
+  struct GlobalUInt64HistogramHandle : GlobalHandle {};
+  struct GlobalDoubleHistogramHandle : GlobalHandle {};
+
+  // Creates instrument in the GlobalInstrumentsRegistry.
+  static GlobalUInt64CounterHandle RegisterUInt64Counter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalDoubleCounterHandle RegisterDoubleCounter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalUInt64HistogramHandle RegisterUInt64Histogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  static GlobalDoubleHistogramHandle RegisterDoubleHistogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, std::vector<absl::string_view> label_keys,
+      std::vector<absl::string_view> optional_label_keys);
+  // Getter functions for StatsPlugins to query registered counters/histograms.
+  static const std::vector<GlobalInstrumentDescriptor>& instruments();
```

Comment:
Note that when a stats plugin is created, it needs a list of all the instrument descriptors so that it can go and create instrument implementation for each descriptor. And, at stats plugin creation time, it doesn't have access to instrument handles.

Reply:
Good point.  But in that case, I would advocate for a `ForEach()`-type interface that lets the stats plugin iterate through all of the registered instruments.  That would accomplish the goal without exposing the internal data structure to all of the stats plugins.